### PR TITLE
Panel select

### PR DESCRIPTION
**Pull request by pykong**
**Up for review and testing.**

This PR updates panel selection based upon caret position in corresponding view.
Thereby the current file is automatically focused and GoTo Next/Prev does not lead to any context switch.

Hence this PR greatly improves user experience.

### CODE REVIEW

**Comment by FichteFoll** [comment]
What feature (set) does this PR add now? Maybe update OP.

File: settings-schema.json [hunk 1]
```diff
@@ -0,0 +1,84 @@
+{
```
Comment (pykong) [inline settings-schema.json]: This shouldn't be part of this PR.

**Comment by braver** [comment]
@pykong please check your PR’s with regards to the notes I put in #751 , each branch should just have the minimum changes related to the feature you’re adding, but you’ve got some other cruft in there as well.

**Comment by braver** [comment]
@pykong can you resolve the merge conflict for me?

**Comment by pykong** [comment]
@braver `on_selection_modified_async` was removed when we switched to the event driven architecture. This led to the new panel update method introduced with this PR never been called. It now works as of: ef701ac

**Comment by braver** [comment]
Ah, makes sense. Thanks!

**Comment by braver** [comment]
I get this really weird behaviour where in util.py it jumps from line 72 to line 69 instead of to the next error at line 83. I can go forwards or back as much as I like but it trips up there for some reason.

weird

**Comment by pykong** [comment]
I can not reproduce this issue. Do you got another example?
Why close this PR?

**Comment by braver** [comment]
I guess it was auto-closed because I removed `next`.

**Comment by pykong** [comment]
panel_select

I am still trying to get any output from `pycodestly`. In lieu of the latter all I can only guess what the described issue is.

I have so far in my testing never encounter a trip up like described. Is it maybe linter related?

**Comment by FichteFoll** [comment]
I noticed in your preview that you have two errors on a line, yet only one is selected in the panel. Other than that this looks very neat.

**Comment by pykong** [comment]
@FichteFoll Thanks. Great you like what you see!
The selection dependends on the column position as well.
But even if more than a single error encompass the column, only the first error will be selected in the panel.

It is a kind of UI question what is best here.

I suggest that we get the error @FichteFoll describes fixed and then merge.
Other features like the folding or changes to the panel selection can follow in their own PRs.

**Comment by FichteFoll** [comment]
>The selection dependends on the column position as well.

Good point. In your particular example, both errors are on the same column though.

I guess the most sensible approach would be to select the nearest error if the caret is between two errors on the same column and hl multiple when they are on the same column.

**Comment by pykong** [comment]
> I guess the most sensible approach would be to select the nearest error if the caret is between two errors on the same column

This is how it is done in the current implementation

> [...] and hl multiple when they are on the same column.

Yes. I think this makes sense to highlight multiple.

**Comment by braver** [comment]
It's broken for everything but the first file in the panel. The second file requires 2 actions to go to the next error, the third file it just jumps back and forth between the file name and the first result.

This is with just hitting "next_result" continuously:

weird

**Comment by pykong** [comment]
I'll check that next week.
Rather busy this WE with other stuff.

File: panel/panel.py [hunk 2]
```diff
@@ -169,10 +196,95 @@ def fill_panel(window, types=None, codes=None, linter=None, update=False):
...
+        colno = None
+
+    full_line = False
+    view_dict = persist.errors.get_view_dict(vid)
+    if not view_dict or util.is_none_or_zero(view_dict["we_count_view"]):
+        return
+
+    line_dicts = view_dict["line_dicts"]
+
+    if lineno:
```
Comment (kaste) [inline panel/panel.py]: What if `lineno == 0`?

File: panel/panel.py [hunk 3]
```diff
@@ -169,10 +196,95 @@ def fill_panel(window, types=None, codes=None, linter=None, update=False):
...
+
+    if not region_dict:
+        return
+
+    if full_line and colno is not None:
+        full_line = region_dict["start"] <= colno <= region_dict["end"]
+
+    panel_lineno = region_dict.get("panel_lineno")
+
+    if panel_lineno:
```
Comment (kaste) [inline panel/panel.py]: The same. Is this really a truthy test or a `None` test?

File: sublime_linter.py [hunk 4]
```diff
@@ -169,6 +169,15 @@ def on_hover(self, view, point, hover_zone):
             if persist.settings.get('show_hover_region_report'):
                 SublimeLinter.shared_plugin().open_tooltip(view, point, True)
 
+    def on_selection_modified_async(self, view):
```
Comment (kaste) [inline sublime_linter.py]: I would really appreciate if we don't immediately start and fill sublime_linter.py with stuff that belongs to a special view? I've shown with the statusbar view and the busy indicator, that this really helps readability. 

I think this is really important especially for open source projects, where a user must find a bug and a way to help and contribute very fast, not knowing the code base.

File: sublime_linter.py [hunk 5]
```diff
@@ -169,6 +169,15 @@ def on_hover(self, view, point, hover_zone):
...
+        if not view:
```
Comment (kaste) [inline sublime_linter.py]: Just curious. Is that really that Sublime calls us sometimes with no view? Just with nothing?

File: panel/panel.py (hunk 3 above)
Comment (kaste) [reply panel/panel.py]: In other words: It is important to state if your data is (0, 0) based and when it is (1, 1) base. (1, 1) based is what we present the user. Bc something like this is known to err very easy.

File: sublime_linter.py (hunk 5 above)
Comment (braver) [reply sublime_linter.py]: I believe because the async-ness the view may no longer exist by the time this code is executed. It’s the reason behind a lot of the `if view` checks in this and other packages.

Comment (FichteFoll) [reply sublime_linter.py]: Indeed, if the async thread was blocking on something earlier while the view was closed (which hopefully doesn't happen but could), then the view would most likely not exist anymore. However, I'm not sure if the function is even called at this point and, in addition, whether the view's ID might just be invalid instead. It's usually safe to check for `view.window()` for when a view object is invalid.

Comment (kaste) [reply sublime_linter.py]: Yeah okay, but you can't just `None` a var that is in our scope. python will not garbage collect this.

So I tested.

```
class UpdateState(sublime_plugin.EventListener):
    def on_selection_modified_async(self, view):
        current_pos = get_current_pos(view)
        import time
        time.sleep(1)
        print('test_view', view.id(), current_pos, view.file_name() or view.name(), view.window())


def get_current_pos(view):
    try:
        return view.rowcol(view.sel()[0].begin())
    except (AttributeError, IndexError):
        return -1, -1

```

If you run this, every callback blocks for a second. So you have plenty of time to close a view. Sublime will then still call you, even after the view is gone. The view is still valid, you can call `id()` and even `rowcol()`, but it is detached so `window()` returns `None`.

File: sublime_linter.py [hunk 6]
```diff
@@ -169,6 +169,15 @@ def on_hover(self, view, point, hover_zone):
...
+            return
+
+        lineno, colno = self.get_line_and_col(view)
```
Comment (kaste) [inline sublime_linter.py]: The `view` here doesn't have to be the `active_view`. 

- You can modify views in the background, and this will trigger.
- If you have multiple views into the same buffer, the view here will always be the *primary* one.

That's why I have the `State` object in `status_bar_view.py`, which holds the `active_view` and `current_pos`.

**Comment by pykong** [comment]
Good points @kaste 
I have moved the call of the `update_panel_selection` function to `status_bar_view`, where it should have been in the first place. Works like a charm.
The function will now simply return early, if `if current_pos == (-1, -1)`, this allowed me to cleanup its logic.

**Comment by braver** [comment]
Ok, seems to work pretty good now. One thing: it doesn't update the selection in the panel as I switch files. Only after I move my cursor does it sync up. Everything else = 🌶

**Comment by braver** [comment]
Please see #797
